### PR TITLE
pc-tty: add framebuffer-based console

### PR DESCRIPTION
**Pull request by adamgreloch**
Adds a graphics mode console that prints terminal output in 8x16 font to a linear framebuffer located under address queried via platformctl/pctl_graphmode in a maximum resolution supported by current graphic mode

This implementation assumes ARGB ordering of the framebuffer

JIRA: RTOS-925

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [X] Tested by hand on: ia32 qemu/hw

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing linter checks and tests passed.
- [X] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [X] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
https://github.com/phoenix-rtos/plo/pull/352
https://github.com/phoenix-rtos/libphoenix/pull/386
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: tty/pc-tty/ttypc.c [hunk 1]
```diff
@@ -132,24 +133,43 @@ int main(void)
 	if ((err = mutexCreate(&ttypc_common.lock)) < 0)
 		return err;
 
+	/* Initialize VGA display */
+	if ((err = ttypc_vga_init(&ttypc_common)) < 0)
+		return err;
```
Comment (agkaminski) [inline tty/pc-tty/ttypc.c:138]: MISRA
```suggestion
	err = ttypc_vga_init(&ttypc_common)
	if (err < 0) {
		return err;
	}
```

File: tty/pc-tty/ttypc.c [hunk 2]
```diff
@@ -132,24 +133,43 @@ int main(void)
...
-		if ((err = ttypc_vt_init(&ttypc_common, _PAGE_SIZE, ttypc_common.vts + i)) < 0)
+		if ((err = ttypc_vt_init(&ttypc_common, ttypc_common.memsz, ttypc_common.vts + i)) < 0)
 			return err;
 	}
 
-	/* Initialize VGA display */
-	if ((err = ttypc_vga_init(&ttypc_common)) < 0)
-		return err;
+	/* Enable fbcon in VTs if available */
+	if (ttypc_common.fbmaxcols != -1 && ttypc_common.fbmaxrows != -1) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc.c:147]: MISRA
```suggestion
	if ((ttypc_common.fbmaxcols != -1) && (ttypc_common.fbmaxrows != -1)) {
```

File: tty/pc-tty/ttypc_fbcon.c [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+ * some weird framebuffers may use something else. In such case it's necessary
+ * to extend pctl_graphmode query with shifts/masks information for each color
+ * from platformctl. */
+#define COLOR_T_TO_RGB32(color) ((0xff << 24) | (color.red << 16) | (color.green << 8) | (color.blue << 0))
+
+
+static void _draw_pixel(ttypc_vt_t *vt, uint16_t x, uint16_t y, uint32_t color)
+{
+	ttypc_t *ttypc = vt->ttypc;
+	*(uint32_t *)((addr_t)ttypc->fbaddr + x * (ttypc->fbbpp / 8) + y * ttypc->fbpitch) = color;
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_fbcon.c:96]: Casting `volatile` away, compiler is free to optimize this away
```suggestion
	*(volatile uint32_t *)((addr_t)ttypc->fbaddr + x * (ttypc->fbbpp / 8) + y * ttypc->fbpitch) = color;
```

File: tty/pc-tty/ttypc_fbcon.c [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+
+void _ttypc_fbcon_drawchar(ttypc_vt_t *vt, int col, int row, uint16_t val)
+{
+	uint8_t i;
+	uint16_t char_pix_y;
+
+	uint8_t fg_color = (val >> 8) & 0xf, bg_color = (val >> 12) & 0xf;
+	const uint32_t arr[] = { COLOR_T_TO_RGB32(colors[fg_color]), COLOR_T_TO_RGB32(colors[bg_color]) };
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_fbcon.c:106]: MISRA nitpick
```suggestion
	const uint32_t arr[2] = { COLOR_T_TO_RGB32(colors[fg_color]), COLOR_T_TO_RGB32(colors[bg_color]) };
```

File: tty/pc-tty/ttypc_fbcon.c [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+{
+	ttypc_t *ttypc = vt->ttypc;
+	*(uint32_t *)((addr_t)ttypc->fbaddr + x * (ttypc->fbbpp / 8) + y * ttypc->fbpitch) = color;
+}
+
+
+void _ttypc_fbcon_drawchar(ttypc_vt_t *vt, int col, int row, uint16_t val)
+{
+	uint8_t i;
+	uint16_t char_pix_y;
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_fbcon.c:103]: Use camelCase
```suggestion
	uint16_t charPixY;
```

File: tty/pc-tty/ttypc_fbcon.c [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+	uint8_t fg_color = (val >> 8) & 0xf, bg_color = (val >> 12) & 0xf;
+	const uint32_t arr[] = { COLOR_T_TO_RGB32(colors[fg_color]), COLOR_T_TO_RGB32(colors[bg_color]) };
+
+	char c = val & 0xff;
+	uint8_t *data = fb_font + TTYPC_FBFONT_BYTES_PER_GLYPH * c;
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_fbcon.c:109]: MISRA nitpick
```suggestion
	uint8_t *data = fb_font + (TTYPC_FBFONT_BYTES_PER_GLYPH * c);
```

File: tty/pc-tty/ttypc_fbfont.h [hunk 7]
```diff
@@ -0,0 +1,4654 @@
+/* The 8x16 font file below is taken from HelenOS, so we're retaining its copyright
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_fbfont.h:1]: Missing Phoenix header

File: tty/pc-tty/ttypc_vga.c [hunk 8]
```diff
@@ -4,7 +4,7 @@
  * VGA display
  *
  * Copyright 2008 Pawel Pisarczyk
- * Copyright 2017, 2012, 2019, 2020 Phoenix Systems
+ * Copyright 2017, 2012, 2019, 2020, 2024 Phoenix Systems
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vga.c:7]: Add yourself to authors too

File: tty/pc-tty/ttypc_vga.c [hunk 9]
```diff
@@ -257,9 +330,23 @@ void _ttypc_vga_getcursor(ttypc_vt_t *vt)
 }
 
 
+#define INVERT_COLORS(val) (((val & 0x0f00) << 4) | ((val & 0xf000) >> 4) | (val & 0xff))
+
+
 void _ttypc_vga_setcursor(ttypc_vt_t *vt)
 {
+	static int curr_pos = -1;
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vga.c:338]: Consider adding this to the `vt` - whole context is held by this structure, this static breaks this rule and makes the whole code non-stateless

File: tty/pc-tty/ttypc_vga.c [hunk 10]
```diff
@@ -289,9 +377,27 @@ int ttypc_vga_init(ttypc_t *ttypc)
...
-	if ((ttypc->vga = mmap(NULL, _PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_PHYSMEM | MAP_ANONYMOUS, -1, (ttypc->color) ? VGA_COLOR : VGA_MONO)) == MAP_FAILED)
-		return -ENOMEM;
+	ttypc->fbmaxcols = -1;
+	ttypc->fbmaxrows = -1;
+
+	if (ttypc_fbcon_init(ttypc) < 0) {
+		/* Map video memory */
+		ttypc->memsz = _PAGE_SIZE;
+		ttypc->vga = mmap(NULL, ttypc->memsz, PROT_READ | PROT_WRITE, MAP_PHYSMEM | MAP_ANONYMOUS, -1, (ttypc->color) ? VGA_COLOR : VGA_MONO);
+		if (ttypc->vga == MAP_FAILED)
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vga.c:387]:
```suggestion
		if (ttypc->vga == MAP_FAILED) {
```

File: tty/pc-tty/ttypc_vga.c [hunk 11]
```diff
@@ -69,39 +73,103 @@ ssize_t _ttypc_vga_read(volatile uint16_t *vga, uint16_t *buff, size_t n)
...
+		for (i = n; i--;) {
 			dvga[i] = svga[i];
+
+			_ttypc_fbcon_drawChar(vt, col, row, svga[i]);
+			if (col > 0) {
+				col--;
+			}
+			else {
+				if (row == 0)
+					break;
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vga.c:168]:
```suggestion
				if (row == 0) {
					break;
				}
```

File: tty/pc-tty/ttypc_vt.c [hunk 12]
```diff
@@ -641,7 +652,30 @@ int ttypc_vt_pollstatus(ttypc_vt_t *vt)
...
+	int ret = EOK;
+
+	switch (cmd) {
+		case FBCONSETMODE:
+			if (vt->fbmode == FBCON_UNSUPPORTED) {
+				ret = -ENODEV;
+				break;
+			}
+			mode = (int)idata;
+			if (mode != FBCON_ENABLED && mode != FBCON_DISABLED) {
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_vt.c:665]:
```suggestion
			if ((mode != FBCON_ENABLED) && (mode != FBCON_DISABLED)) {
```

File: tty/pc-tty/ttypc_fbcon.c [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+ * some weird framebuffers may use something else. In such case it's necessary
+ * to extend pctl_graphmode query with shifts/masks information for each color
+ * from platformctl. */
+#define COLOR_T_TO_RGB32(color) (((((uint32_t)0xffU) << 24U) | (((uint32_t)(color).red) << 16U) | (((uint32_t)(color).green) << 8U) | ((uint32_t)(color).blue)))
+
+
+static void _drawPixel(ttypc_vt_t *vt, uint16_t x, uint16_t y, uint32_t color)
+{
+	ttypc_t *ttypc = vt->ttypc;
+	*(volatile uint32_t *)((addr_t)ttypc->fbaddr + (x * (ttypc->fbbpp / 8U)) + (y * ttypc->fbpitch)) = color;
```
[the remark this answers is not in the document]
Comment (adamgreloch) [reply tty/pc-tty/ttypc_fbcon.c]: Changed to `uintptr_t` for now. What's the idea behind `char *`?

[the remark this answers is not in the document]
Comment (agkaminski) [reply tty/pc-tty/ttypc_fbcon.c]: char * is industry standard for pointer arithmetic. Going to plain integers (like addr_t or uintptr_t) is a last resort (when we e.g. start with integers, like device base address or something like that). In this case we start with pointer and end with pointer, so it's best to keep this in a pointer realm

[the remark this answers is not in the document]
Comment (adamgreloch) [reply tty/pc-tty/ttypc_fbcon.c]: Right, valid point. Corrected

File: tty/pc-tty/ttypc.c [hunk 14]
```diff
@@ -132,24 +133,44 @@ int main(void)
...
 
+	/* Initialize VGA display */
+	err = ttypc_vga_init(&ttypc_common);
+	if (err < 0)
+		return err;
+
 	/* Initialize VTs */
 	for (i = 0; i < NVTS; i++) {
-		if ((err = ttypc_vt_init(&ttypc_common, _PAGE_SIZE, ttypc_common.vts + i)) < 0)
+		if ((err = ttypc_vt_init(&ttypc_common, ttypc_common.memsz, ttypc_common.vts + i)) < 0)
```
[the remark this answers is not in the document]
Comment (adamgreloch) [reply tty/pc-tty/ttypc.c]: Uuuh, apologies. No idea what happened here. Should be fixed now

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**